### ISSUE TITLE
options not available in configure function in test_package

### ISSUE THREAD

**Issue by bresniv** [conan-io/conan#1560]
The passed options can't be read from the configure function when running a test package.

Using the [this package](https://gist.github.com/bresniv/176aec9a9374d335bdb1d559ea4d8124), this set of commands was executed:
```
conan export user/channel -o shared=True
conan install conan_package/0.1.145@user/channel --build=conan_package -o shared=True
conan test_package user/channel --test-only -o shared=True
```

After which the following warning was shown:
```
conan_package/0.1.145@user/channel test package: WARN: Can't find a 'conan_package/0.1.145@user/channel' package for the specified options and settings:
- Settings:
- Options: shared=False

ERROR: Missing prebuilt package for 'conan_package/0.1.145@user/channel'
Try to build it from sources with "--build conan_package"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```

It seems the option shared is reset to false between the requirements and configure function (output test_package):
```
Package shared (requirements): True conan_test_package
Package shared (configure): False conan_test_package
conan_package False
Package shared (requirements): False conan_test_package
Package shared (configure): False conan_package
Package shared (requirements): False conan_package
```

Is this a bug introduced in 0.25 or am I using something wrong?

**Comment by memsharded** [comment]
I think you might be missing the package option specifier:

```bash
$ conan install conan_package/....@.../... -o conan_package:shared=True
```

Could you please try and tell me if it worked?

Other tips:
- ``conan export`` does not use options
-  you might be interested in the ``conan create`` command, it basically does what you are doing, but with one command.

**Comment by bresniv** [comment]
The problem is not in the ```conan install``` command as this works fine.
The problem is with running the ```conan test_package```.
When ```conan test_package user/channel --test-only -o shared=True``` is run, the ```shared``` option is still ```True``` when read in the ```requirements``` function but is ```False``` when read in the ```configure``` function.

This problem does not seen to occur when using ```conan create -o shared=True user/channel```. But using ```conan create``` is not possible in our situation.

**Comment by memsharded** [comment]
I think you need to use the package specifier, like ``-o package_name:shared=True``. Your line seems to miss it, did you try it? In some cases, the bare use without the package name is supported for legacy reasons, but the complete one is the correct one.

**Comment by bresniv** [comment]
```conan test_package --test-only -o conan_test_package:shared=True user/channel``` provides the same result:
```
Package shared (requirements): True conan_test_package
Package shared (configure): False conan_test_package
Package shared (requirements): False conan_test_package
Package shared (configure): False conan_package
Package shared (requirements): False conan_package
conan_package/0.1.145@user/channel test package: Installing dependencies
conan_package/0.1.145@user/channel test package: WARN: Can't find a 'conan_package/0.1.145@user/channel' package for the specified options and settings:
- Settings:
- Options: shared=False

ERROR: Missing prebuilt package for 'conan_package/0.1.145@user/channel'
Try to build it from sources with "--build conan_package"
Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
```
option ```shared``` can be used until the first time the ```requirements``` function is called in the ```conan_test_package``` after that it changes to ```False``` for some reason

commands used used on [these files](https://gist.github.com/bresniv/176aec9a9374d335bdb1d559ea4d8124):
```
$ conan export user/channel
$ conan install conan_package/0.1.145@user/channel --build=conan_package -o conan_package:shared=True
$ conan test_package --test-only -o conan_test_package:shared=True user/channel
```

ps:
I did not now omitting the package name for options was a legacy function, how does this work with test packages as these don't need to have a name?

**Comment by memsharded** [comment]
Thanks very much for the details. Indeed it seems a bug, or at least some very weird behavior.
I am investigating it.

**Comment by memsharded** [comment]
Confirmed, it was a subtle bug, because of someone modifying an object passed by reference, and destroying the data that is necessary later. A pull requests https://github.com/conan-io/conan/pull/1566 has been submitted, to be in next release.

Note that for the new ``conan create`` command, it has to be specified the package option not the test_package option, like:

```bash
$ create lasote/stable -o conan_package:shared=True -o conan_test_package:shared=True
```

This is because the flow is different. First the package is exported + installed, but "test_package" is not used at all. Finally, if "test_package" folder exists, it will be used, but it is assumed that it will use the same configuration as the created package. In your case, as the test_package is changing the package options, it might be inconsistent.

While the fix is released, and in fact as possible best practice I would try:

- Define and use the option in the "package", not in the "test_package"
- This is more robust, as some test_package requirement might not have a ``shared`` option
- Install with ``-o conan_package:shared=True``
- When you need the package option inside test_package, you can just read it with ``self.options["conan_package"].shared``

**Comment by memsharded** [comment]
Fixed in develop, will be released in 0.26

**Comment by bresniv** [comment]
Thanks, looking forward to version 0.26

**Comment by memsharded** [comment]
Released in 0.26, please try
